cedar-policy/src/ffi/format.rs

```diff
@@ -0,0 +1,164 @@
+
+#![allow(clippy::module_name_repetitions)]
+
+use super::utils::DetailedError;
+use cedar_policy_formatter::{policies_str_to_pretty, Config};
+use serde::{Deserialize, Serialize};
+#[cfg(feature = "wasm")]
+use wasm_bindgen::prelude::wasm_bindgen;
+
+#[cfg(feature = "wasm")]
+extern crate tsify;
+
+/// Apply the Cedar policy formatter to a policy set in the Cedar policy format
+#[cfg_attr(feature = "wasm", wasm_bindgen(js_name = "formatPolicies"))]
+#[allow(clippy::needless_pass_by_value)]
+pub fn format(call: FormattingCall) -> FormattingAnswer {
+    let config = Config {
+        line_width: call.line_width,
+        indent_width: call.indent_width,
+    };
+    match policies_str_to_pretty(&call.policy_text, &config) {
+        Ok(prettified_policy) => FormattingAnswer::Success {
+            formatted_policy: prettified_policy,
+        },
+        Err(err) => FormattingAnswer::Failure {
+            errors: vec![err.into()],
+        },
+    }
+}
+
+/// Apply the Cedar policy formatter. Input is a JSON encoding of
+/// [`FormattingCall`] and output is a JSON encoding of [`FormattingAnswer`].
+///
+/// # Errors
+///
+/// Will return `Err` if the input JSON cannot be deserialized as a
+/// [`FormattingCall`].
+pub fn format_json(json: serde_json::Value) -> Result<serde_json::Value, serde_json::Error> {
+    let ans = format(serde_json::from_value(json)?);
+    serde_json::to_value(ans)
+}
+
+/// Apply the Cedar policy formatter. Input and output are strings containing
+/// serialized JSON, in the shapes expected by [`format_json()`].
+///
+/// # Errors
+///
+/// Will return `Err` if the input cannot be converted to valid JSON or
+/// deserialized as a [`FormattingCall`].
+pub fn format_json_str(json: &str) -> Result<String, serde_json::Error> {
+    let ans = format(serde_json::from_str(json)?);
+    serde_json::to_string(&ans)
+}
+
+/// Struct containing the input data for formatting
+#[derive(Serialize, Deserialize, Debug)]
+#[cfg_attr(feature = "wasm", derive(tsify::Tsify))]
+#[cfg_attr(feature = "wasm", tsify(into_wasm_abi, from_wasm_abi))]
+#[serde(rename_all = "camelCase")]
+#[serde(deny_unknown_fields)]
+pub struct FormattingCall {
+    /// Policy text. May define multiple policies or templates in the Cedar policy format.
+    policy_text: String,
+    /// Line width
+    line_width: usize,
+    /// Indentation width
+    indent_width: isize,
+}
+
+/// Result struct for formatting
+#[derive(Debug, Serialize, Deserialize)]
+#[cfg_attr(feature = "wasm", derive(tsify::Tsify))]
+#[cfg_attr(feature = "wasm", tsify(into_wasm_abi, from_wasm_abi))]
+#[serde(tag = "type")]
+#[serde(rename_all = "camelCase")]
+pub enum FormattingAnswer {
+    /// Represents a failure to call the formatter
+    Failure {
+        /// Policy parse errors
+        errors: Vec<DetailedError>,
+    },
+    /// Represents a successful formatting call
+    Success {
+        /// Formatted policy
+        formatted_policy: String,
+    },
```

Comment:
Any reason we have `Failure` first in this enum, and `Success` first in the others?

Reply:
No reason. Does it matter?

Reply:
It does not